File: src/main/java/com/zippeykeys/praisebe/block/base/BlockIdolBase.java
package com.zippeykeys.praisebe.block.base;

import com.zippeykeys.praisebe.block.tile.base.TileEntityIdolBase;

import org.jetbrains.annotations.Nullable;

import lombok.val;
import lombok.var;

import net.minecraft.block.BlockHorizontal;
import net.minecraft.block.SoundType;
import net.minecraft.block.material.Material;
import net.minecraft.block.properties.PropertyDirection;
import net.minecraft.block.state.BlockStateContainer;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.EnumBlockRenderType;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;


public class BlockIdolBase extends PBBlock{
    public static final PropertyDirection FACING = BlockHorizontal.FACING;

    public BlockIdolBase(String nameIn, Material materialIn){
        super(nameIn, materialIn);
        setHardness(1.5f);
        setResistance(6000000.0F);
        setHarvestLevel("pickaxe", 1);
        setDefaultState(blockState.getBaseState()
          .withProperty(FACING, EnumFacing.NORTH));
        setSoundType(SoundType.STONE);
    }

    @Override
    public IBlockState getStateForPlacement(World world, BlockPos pos, EnumFacing facing, float hitX, float hitY,
      float hitZ, int meta, EntityLivingBase placer, EnumHand hand){
        return getDefaultState().withProperty(FACING, facing);
    }

    @Override
    public IBlockState getActualState(IBlockState state, IBlockAccess worldIn, BlockPos pos){
        var facing = EnumFacing.NORTH;
        val tile = worldIn.getTileEntity(pos);
        val tileClazz = getTileEntity();
        if(tileClazz != null && tileClazz.isInstance(tile)){
            facing = EnumFacing.getFront(tileClazz.cast(tile)
              .getFacing());
        }
        return state.withProperty(FACING, facing);
    }

    @Override
    public IBlockState getStateFromMeta(int meta){
        var enumFacing = EnumFacing.getFront(meta);
        if(enumFacing.getAxis() == EnumFacing.Axis.Y){
            enumFacing = EnumFacing.NORTH;
        }
        return getDefaultState().withProperty(FACING, enumFacing);
    }

    @Override
    public int getMetaFromState(IBlockState state){
        return state.getValue(FACING)
          .getIndex();
    }

    @Override
    public int damageDropped(IBlockState state){
        return getMetaFromState(state);
    }

    @Override
    public boolean isNormalCube(IBlockState state, IBlockAccess world, BlockPos pos){
        return false;
    }

    @Override
    public boolean isOpaqueCube(IBlockState state){
        return false;
    }

    @Override
    public boolean isFullCube(IBlockState state){
        return false;
    }

    @Override
    public EnumBlockRenderType getRenderType(IBlockState state){
        return EnumBlockRenderType.MODEL;
    }

    @Override
    public boolean onBlockActivated(World worldIn, BlockPos pos, IBlockState state, EntityPlayer playerIn,
      EnumHand hand, EnumFacing facing, float hitX, float hitY, float hitZ){
        val self = worldIn.getTileEntity(pos);
        if(self == null || playerIn.isSneaking()){ return false; }
        worldIn.notifyBlockUpdate(pos, state, state, 3);
        return true;
    }

    @Override
    protected BlockStateContainer createBlockState(){
        return new BlockStateContainer.Builder(this) //
          .add(FACING) //
          .build();
    }

    @Override
    public @Nullable Class<? extends TileEntityIdolBase> getTileEntity(){
        return null;
    }
}

File: src/main/kotlin/com/zippeykeys/praisebe/block/base/BlockIdolBase.kt
package com.zippeykeys.praisebe.block.base

import com.zippeykeys.praisebe.block.tile.base.TileEntityIdolBase
import net.minecraft.block.BlockHorizontal
import net.minecraft.block.SoundType
import net.minecraft.block.material.Material
import net.minecraft.block.state.BlockStateContainer
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.EnumBlockRenderType
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraft.world.IBlockAccess
import net.minecraft.world.World


class BlockIdolBase(nameIn: String, materialIn: Material) : PBBlock(nameIn, materialIn) {
    override val tileEntity: Class<out TileEntityIdolBase>?
        get() = null

    init {
        setHardness(1.5f)
        setResistance(6000000.0f)
        setHarvestLevel("pickaxe", 1)
        defaultState = blockState.baseState
                .withProperty(FACING, EnumFacing.NORTH)
        soundType = SoundType.STONE
    }

    override fun getStateForPlacement(world: World, pos: BlockPos, facing: EnumFacing, hitX: Float, hitY: Float,
            hitZ: Float, meta: Int, placer: EntityLivingBase, hand: EnumHand): IBlockState {
        return defaultState.withProperty(FACING, facing)
    }

    @Suppress("OverridingDeprecatedMember")
    override fun getActualState(state: IBlockState, worldIn: IBlockAccess, pos: BlockPos): IBlockState {
        var facing = EnumFacing.NORTH
        val tile = worldIn.getTileEntity(pos)
        val tileClazz = tileEntity
        if (tileClazz != null && tileClazz.isInstance(tile)) {
            facing = EnumFacing.getFront(tileClazz.cast(tile)
                    .facing)
        }
        return state.withProperty(FACING, facing)
    }

    @Suppress("OverridingDeprecatedMember")
    override fun getStateFromMeta(meta: Int): IBlockState {
        var enumFacing = EnumFacing.getFront(meta)
        if (enumFacing.axis == EnumFacing.Axis.Y) {
            enumFacing = EnumFacing.NORTH
        }
        return defaultState.withProperty(FACING, enumFacing)
    }

    override fun getMetaFromState(state: IBlockState): Int {
        return state.getValue(FACING)
                .index
    }

    override fun damageDropped(state: IBlockState): Int {
        return getMetaFromState(state)
    }

    override fun isNormalCube(state: IBlockState, world: IBlockAccess, pos: BlockPos): Boolean {
        return false
    }

    @Suppress("OverridingDeprecatedMember")
    override fun isOpaqueCube(state: IBlockState): Boolean {
        return false
    }

    @Suppress("OverridingDeprecatedMember")
    override fun isFullCube(state: IBlockState): Boolean {
        return false
    }

    @Suppress("OverridingDeprecatedMember")
    override fun getRenderType(state: IBlockState): EnumBlockRenderType {
        return EnumBlockRenderType.MODEL
    }

    override fun onBlockActivated(worldIn: World, pos: BlockPos, state: IBlockState, playerIn: EntityPlayer,
            hand: EnumHand, facing: EnumFacing, hitX: Float, hitY: Float, hitZ: Float): Boolean {
        val self = worldIn.getTileEntity(pos)
        if (self == null || playerIn.isSneaking) {
            return false
        }
        worldIn.notifyBlockUpdate(pos, state, state, 3)
        return true
    }

    override fun createBlockState(): BlockStateContainer {
        return BlockStateContainer.Builder(this)
                .add(FACING)
                .build()
    }

    companion object {
        val FACING = BlockHorizontal.FACING
    }
}